crates/rolldown_binding/src/types/binding_outputs.rs

```
@@ -65,41 +67,46 @@ impl From<Vec<rolldown_common::Output>> for BindingOutputs {
 }
 
-pub fn update_outputs(
-  outputs: &mut Vec<rolldown_common::Output>,
-  changed: JsChangedOutputs,
-) -> anyhow::Result<()> {
-  for chunk in changed.chunks {
-    if let Some(index) = outputs.iter().position(|o| match o {
-      rolldown_common::Output::Chunk(v) => v.preliminary_filename == chunk.preliminary_filename,
-      rolldown_common::Output::Asset(_) => false,
-    }) {
-      match &mut outputs[index] {
-        rolldown_common::Output::Chunk(old_chunk) => {
-          update_output_chunk(old_chunk, chunk)?;
+impl JsChangedOutputs {
+  pub fn apply_changes(
+    &mut self,
+    outputs: &mut Vec<rolldown_common::Output>,
+  ) -> anyhow::Result<()> {
+    let mut result = Ok(());
+    if !self.deleted.is_empty() || !self.changes.is_empty() {
+      outputs.retain_mut(|output| {
+        if result.is_err() {
+          return true;
         }
-        rolldown_common::Output::Asset(_) => {}
-      }
-    }
-  }
-  for asset in changed.assets {
-    if let Some(index) = outputs.iter().position(|o| o.filename() == asset.filename) {
-      outputs[index] = rolldown_common::Output::Asset(Arc::new(asset.into()));
-    }
-  }
-  for deleted in changed.deleted {
-    if let Some(index) = outputs.iter().position(|o| o.filename() == deleted) {
-      outputs.remove(index);
+        let filename = output.filename();
+        if self.deleted.contains(filename) {
+          return false;
+        }
+        if let Some(change) = self.changes.remove(filename) {
+          match (output, change) {
+            (rolldown_common::Output::Chunk(old_chunk), Either::A(chunk)) => {
+              if let Err(err) = update_output_chunk(old_chunk, chunk) {
+                result = Err(err);
+              }
+            }
+            (v @ rolldown_common::Output::Asset(_), Either::B(asset)) => {
+              *v = rolldown_common::Output::Asset(Arc::new(asset.into()));
+            }
+            _ => {}
```

I think we can return an error (`result = Err(...)`) here that changing chunk to asset or asset to chunk is not supported.

---

Okay, let me improve it later.